### ISSUE TITLE
scalatest 2.12 runs 7x slower than without?

### ISSUE THREAD

**Issue by dvardy** [scalatest/scalatest#2175]
We have a test framework written in scala, inside intelliJ using Maven. We've taken one "testcase" and implemented it as a scalatest test. And removed all other testcases from the project. This served as a POC.

Our maven pom utilizes: 
<scala.version>2.12.12</scala.version>

    <dependency>
      <groupId>org.scalatest</groupId>
      <artifactId>scalatest_2.12</artifactId>
      <version>3.0.8</version>
    </dependency>

The scalatest version takes 7x longer to run (via the advanced scalatest runner), which isn't tenable. Is this to be expected? Or is there something we should ensure to do to speed up execution?

When comparing the timestamps in the logs, apart from printlns, it looks like commands are all taking longer when compared between the two runs. (such that the average test length is 7x more)

Each of our two runs were issued on a databricks cluster:
8 Workers
448 GB Memory
128 Cores
1 Driver
14 GB Memory, 4 Cores
Runtime
7.3.x-scala2.12

via a databricks job.

**non-scalatest project with databricks job details:**
Main class: uk.xxx.xxx.xxx.Main
Dependent libraries: (a single jar containing our non scalatest project)
Parameters: ["x","y"]
Time to run: **5 mins**

**scalatest with databricks job details:**
Main class: org.scalatest.tools.Runner 
Dependent libraries: (a single jar containing our scalatest project)
Parameters: ["-R","dbfs:/ourFatJar.jar","-o","-f","/dbfs/ourlog.txt","-s","uk.xxx.xxx.xxx.Main","-DmaxConnections=100"]
Time to run: **34 mins**

**Comment by cheeseng** [comment]
@dvardy Is it possible you can share a minimal setup for both case that we can run and inspect the cause?  It will be better also if the comparison is done with with the latest version of ScalaTest, which is 3.2.14, we do not have plan to release new minor version for 3.0.x.

From what you provided, it seems both are running uk.xxx.xxx.xxx.Main, which I am not sure what it is doing, I guess it is written as a ScalaTest Suite via its lifecycle overriden?

**Comment by dvardy** [comment]
@cheeseng Thanks for your response. We are looking now at trying to run our poc test with 3.2.14 as suggested, to see first if we still hit the performance issue.

Could you confirm, as per the sample code at:
https://www.scalatest.org/scaladoc/3.2.14/org/scalatest/featurespec/AnyFeatureSpec.html

Which maven libraries (ideally as coordinates, for scala 2.12 and scalatest 3.2.14) and imports are needed? (just to run that exact sample code)

The sample code only mentions a single import of org.scalatest._
However, with only the one import, we get an error of: not found: value assertionsHelper

Thanks

**Comment by cheeseng** [comment]
@dvardy import org.scalatest._ should be fine, afaik assertionsHelper is from the code generated by the assert macro, as long as you are in the scope of org.scalatest.Assertions subclass (e.g. AnyFeatureSpec), you should be fine.

It is getting a bit late here, I'll call it a day soon.  In case you still encounter problem to use 3.2.14 feel free to share a minimal project that encounter the problem, I can help take a look and see if I can help tomorrow.

Cheers.

**Comment by dvardy** [comment]
@cheeseng Thanks. hmm. I was initially only trying in databricks, with:

```
import org.scalatest._

// import org.scalatest.featurespec._
// import org.scalatest.shouldmatchers._
// import org.scalatest.matchers-core._
// import org.scalatest.core._
// import org.scalatest.compatible._
// import org.scalatest.Assertions._

class myClass extends featurespec.AnyFeatureSpec with GivenWhenThen {
  info("As...")

  Feature("myfeature") {
    Scenario("myscenario") {
      Given("given...")
      When("when...")
      Then("then...")
      assert(1==1)
    }
  }
}

run(new myClass)
```

which was throwing the error. I tried loading many libraries into the cluster (removing the default 3.0.8 first, then re-attaching notebook):

![image](https://user-images.githubusercontent.com/116662500/198084006-5354aed4-3d54-4113-8218-6a80018007a9.png)

However, going to our project in intelliJ and setting only the one scalatest library in the pom, certainly works locally. Perhaps there's still some residue of scalatest 3.0.8 in the cluster that I've not efficiently removed, causing issues.

We'll attempt to run our 3.2.14 project jar in databricks via databricks job with the high spec 11.3 cluster, and compare it with speed of running without scalatest implementation, on same 11.3 cluster. Hoping there's not much of a performance difference this time.

**Comment by cheeseng** [comment]
@dvardy Your code seems to build and run fine with this sample sbt project here: 

https://github.com/cheeseng/test-databricks

It would be great to get scalatest running fine in databricks, to investigate the cause I'll need to be able to reproduce the problem, do let me know if there's anything I can help.

Cheers.

**Comment by cheeseng** [comment]
> @cheeseng We managed to get our test project working using 3.2.14 (we had to ensure that 3.0.8 included with databricks runtime was cleared out successfully within an init script).
> 
> However, there's only a mild performance increase with 3.2.14 over 3.0.8
> 
> Without scalatest: 5.5mins With scalaTest 3.0.8: 34 mins With scalaTest 3.2.14: 26 mins (all on a 7.3 cluster)
> 
> Unfortunately we can't hand the project over for you to test.
> 
> We're left with some possible options to assess other performance scenarios:
> 
> 1. Try on an 11.3 cluster (once we can configure spark correctly)
> 2. Try not using the scalaTest advanced runner, but using the scalatest-maven-plugin, either running on the cluster something (through init script possible) or on the devops build agent
> 3. Come up with some general operations that we can include in a sample project that show difference in speed between scalatest and without, that we can submit here.
> 4. Try an alternative to scalaTest like Junit

@dvardy I'll be happy to investigate if you can provide 3., even if it involves some setup for spark.  And yes the way you're running ScalaTest is a bit different than the common way (e.g. through sbt or scalatest-maven-plugin), one thing is parallel execution in scalatest is achieved by mixing ParallelTestExecution trait (to run tests in a Suite in parallel) or run multiple suites in parallel by passing in -P[S] to org.scalatest.tools.Runner.

**Comment by dvardy** [comment]
@cheeseng We're looking at scalatest performance when running from inside our project. What would the appropriate arguments be to run a sample test with the runner? For example, given:

```
Main.scala file:

import org.scalatest.tools.Runner

object Main {
  def main { 
      var myargs = Array("-s uk.xx.xx.xx.testSuite")
      Runner.main(myargs)
  }
}

class testSuite {
  info("As...")
  
    Feature("myfeature") {
      Scenario("myscenario") {
        Given("given...")
        When("when...")
        Then("then...")
        assert(1==1)
      }
    }
}

//end of Main.scala file
```

**Comment by dvardy** [comment]
@cheeseng Thanks, yep, that works. Trying it on a databricks cluster presently.

**Comment by dvardy** [comment]
@cheeseng The testcase looks to have run successfully (succeeded: 1) and with good performance. Though we haven't yet set it to output to file for any log or results, trying that now to see if there's a performance decrease.

This is probably a separate issue, but despite a testcase succeeding, we still get error:


```
com.databricks.backend.daemon.driver.ExitSecurityException: Program attempted to exit with code 0
at com.databricks.backend.daemon.driver.TrapExitSecurityManager.checkExit(DriverLocal.scala:960)
	at java.lang.Runtime.exit(Runtime.java:108)
	at java.lang.System.exit(System.java:973)
	at org.scalatest.tools.Runner$.main(Runner.scala:778)
```

Any idea how we can exit successfully (if all scalatests pass) after invoking the runner (from within our Main class) as in:

object Main {
  def main {
      var myargs = Array("-s", "x.x.x.x.testSuite", "-o")
      Runner.main(myargs)
  }
}

class testSuite {
info, feature, scenario...
}

**Comment by cheeseng** [comment]
@dvardy Hmm, I think the main function is calling System.exit(0) here at line 778 (as hinted from your stack trace): 

https://github.com/scalatest/scalatest/blob/main/jvm/core/src/main/scala/org/scalatest/tools/Runner.scala#L778

I think that is having some issue with databricks TrapExitSecurityManager?  Perhaps we can try add an option or system property to skip calling System.exit(0) and just let it complete the main function.

Or you can call run instead of main: 

https://github.com/scalatest/scalatest/blob/main/jvm/core/src/main/scala/org/scalatest/tools/Runner.scala#L793

Hope this helps.

Cheers.

**Comment by dvardy** [comment]
Brilliant, thanks @cheeseng much appreciated for all your help!
